File: countPaths.py
MOD = 1_000_000_007

def dfs(s, adj, vis, v):
    if vis[s]:
        return
    vis[s] = True
    for i in adj[s]:
        dfs(i, adj, vis, v)
    v.append(s)

def solve():
    import sys
    input = sys.stdin.read
    data = input().split()
    
    index = 0
    n = int(data[index])
    index += 1
    m = int(data[index])
    index += 1

    adj = [[] for _ in range(n + 1)]  # Adjust size based on n
    vis = [False] * (n + 1)
    v = []

    for _ in range(m):
        a = int(data[index])
        index += 1
        b = int(data[index])
        index += 1
        adj[b].append(a)

    for i in range(1, n + 1):
        if not vis[i]:
            dfs(i, adj, vis, v)

    t = []
    i, f = 0, 0
    while i < len(v):
        if f == 0:
            if v[i] == 1:
                f = 1
                t.append(1)
        else:
            t.append(v[i])
        i += 1

    path = [0] * (n + 1)
    path[1] = 1
    for x in t:
        for j in adj[x]:
            path[x] = (path[x] + path[j]) % MOD

    print(path[n])

def main():
    t = 1
    # t = int(input())
    for _ in range(t):
        solve()
        # print()

if __name__ == "__main__":
    main()

File: countPaths.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    static final int MOD = 1_000_000_007;
    static List<Integer>[] adj = new ArrayList[100005];
    static boolean[] vis = new boolean[100005];
    static List<Integer> v = new ArrayList<>();

    static void dfs(int s) {
        if (vis[s]) return;
        vis[s] = true;
        for (int i : adj[s]) {
            dfs(i);
        }
        v.add(s);
    }

    static void solve(Scanner scanner) {
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        
        for (int i = 0; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            int a = scanner.nextInt();
            int b = scanner.nextInt();
            adj[b].add(a);
        }

        for (int i = 1; i <= n; i++) {
            if (!vis[i]) {
                dfs(i);
            }
        }

        List<Integer> t = new ArrayList<>();
        int i = 0, f = 0;
        while (i < v.size()) {
            if (f == 0) {
                if (v.get(i) == 1) {
                    f = 1;
                    t.add(1);
                }
            } else {
                t.add(v.get(i));
            }
            i++;
        }

        long[] path = new long[100005];
        path[1] = 1;
        for (int x : t) {
            for (int j : adj[x]) {
                path[x] = (path[x] + path[j]) % MOD;
            }
        }

        System.out.println(path[n]);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = 1;
        // t = scanner.nextInt();
        while (t-- > 0) {
            solve(scanner);
            // System.out.println();
        }
    }
}
